### PR TITLE
[Rework] ExprInterface.getType()

### PR DESCRIPTION
**Pull request by xeren**
Introduces the `Type` class, which extends the `IExpr.getPrecision()` property.  It will replace it  Currently, the supported types are `BooleanType` and `IntegerType`.  Due to name conflicts, the `INonDetTypes` class is also replaced here.

### CODE REVIEW

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/boogie/SvcompProcedures.java [hunk 1]
```diff
@@ -97,45 +100,41 @@ public static void __VERIFIER_atomic_end(VisitorBoogie visitor) {
...
+		String suffix = name.substring("__VERIFIER_nondet_".length());
+		boolean signed = switch (suffix) {
+			case "int", "short", "long", "char" -> true;
+			default -> false;
+		};
+		BigInteger min = switch (suffix) {
+			case "long" -> BigInteger.valueOf(Long.MIN_VALUE);
+			case "int" -> BigInteger.valueOf(Integer.MIN_VALUE);
+			case "short" -> BigInteger.valueOf(Short.MIN_VALUE);
+			case "char" -> BigInteger.valueOf(-128);
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/boogie/SvcompProcedures.java:117]: You can use `Byte.MIN_VALUE`

File: dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/boogie/SvcompProcedures.java [hunk 2]
```diff
@@ -97,45 +100,41 @@ public static void __VERIFIER_atomic_end(VisitorBoogie visitor) {
...
+			default -> BigInteger.ZERO;
+		};
+		BigInteger max = switch (suffix) {
+			case "int" -> BigInteger.valueOf(Integer.MAX_VALUE);
+			case "uint", "unsigned_int" -> UnsignedInteger.MAX_VALUE.bigIntegerValue();
+			case "short" -> BigInteger.valueOf(Short.MAX_VALUE);
+			case "ushort", "unsigned_short" -> BigInteger.valueOf(65535);
+			case "long" -> BigInteger.valueOf(Long.MAX_VALUE);
+			case "ulong" -> UnsignedLong.MAX_VALUE.bigIntegerValue();
+			case "char" -> BigInteger.valueOf(127);
```
Comment (ThomasHaas) [inline dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/boogie/SvcompProcedures.java:127]: `Byte.MAX_VALUE`

File: dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExprInterface.java [hunk 3]
```diff
@@ -1,17 +1,22 @@
...
+import com.dat3m.dartagnan.expression.type.Type;
+import com.dat3m.dartagnan.expression.type.TypeFactory;
 import com.dat3m.dartagnan.program.Register;
 import com.google.common.collect.ImmutableSet;
 
 public interface ExprInterface {
 
+    default Type getType() {
+        return TypeFactory.getInstance().getBooleanType();
+    }
```
[the remark this answers is not in the document]
Comment (xeren) [reply dartagnan/src/main/java/com/dat3m/dartagnan/expression/ExprInterface.java:13]: That was a case of laziness.

**Review by ThomasHaas — APPROVED** [review]
LGTM. I will merge once the CI is done

**Pull request merged**